### PR TITLE
feat: implement `ecadd` `ecmul`  `ecpairing` precompiles

### LINKED ISSUE #343: Implement `ecAdd` precompile (0x06)
More info in [evm.codes/precompiled](https://www.evm.codes/precompiled)

### LINKED ISSUE #344: Implement `ecMul` precompile (0x07)
More info in [evm.codes/precompiled](https://www.evm.codes/precompiled)

### LINKED ISSUE #345: Implement `ecPairing` precompile (0x08)
More info in [evm.codes/precompiled](https://www.evm.codes/precompiled)

### PR DESCRIPTION
**Pull request by TomasArrachea**
Closes #343
Closes #344
Closes #345 

This PR adds the rust implementation for the ecadd, ecmul, ecpairing precompiles.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -23,6 +23,8 @@ bitflags = "2.5.0"
 secp256k1 = { version = "0.29.0", features = ["recovery"] }
 sha2 = "0.10.8"
 ripemd = "0.1.3"
+substrate-bn = "0.6.0"
```
Comment (MegaRedHand) [inline Cargo.toml:26]: We can now use lambdaworks for the pairing, so we can remove this dep

Comment (juanimedone) [reply Cargo.toml:26]: On it!

**Comment by juanimedone** [comment]
The three `ecadd`, `ecmul` and `ecpairing` precompiles are now implemented with `lambdaworks`. All of them have unit tests contemplating edge cases.

File: src/precompiles.rs [hunk 2]
```diff
@@ -124,6 +142,156 @@ pub fn modexp(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes
...
+
+    // (0,0) represents infinity, in that case the other point (if valid) should be directly returned
+    let zero_el = BN254FieldElement::from(0);
+    let p1_is_infinity = x1.eq(&zero_el) && y1.eq(&zero_el);
+    let p2_is_infinity = x2.eq(&zero_el) && y2.eq(&zero_el);
+
+    match (p1_is_infinity, p2_is_infinity) {
+        (true, true) => return Bytes::from([0u8; 64].to_vec()),
+        (true, false) => {
+            if let Ok(p2) = BN254Curve::create_point_from_affine(x2, y2) {
```
Comment (belfortep) [inline src/precompiles.rs:166]: I think is just a naming thing, but the evm precompiles calls for a `bn128` curve and here we create a `bn254`

File: src/precompiles.rs [hunk 3]
```diff
@@ -124,6 +142,156 @@ pub fn modexp(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes
...
+    if let Ok(p) = BN254Curve::create_point_from_affine(x1, y1) {
+        let mul = p.operate_with_self(s).to_affine();
+        let res = [mul.x().to_bytes_be(), mul.y().to_bytes_be()].concat();
+        return Bytes::from(res);
+    }
+    Bytes::new()
+}
+
+pub fn ecpairing(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes {
+    let gas_cost = ECPAIRING_COST + (34_000 * (calldata.len() as u64 / 192));
```
Comment (belfortep) [inline src/precompiles.rs:226]: We could move the `34_000` number to a constant

File: src/precompiles.rs [hunk 4]
```diff
@@ -124,6 +142,156 @@ pub fn modexp(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes
...
+        let pairing_result = match BN254AtePairing::compute_batch(&[(&p1, &p2)]) {
+            Ok(result) => result,
+            Err(_) => return Bytes::from([0u8; 32].to_vec()),
+        };
+        mul *= pairing_result;
+    }
+
+    let success = mul.eq(&QuadraticExtensionFieldElement::one());
+    let mut output = vec![0_u8; 32];
+    output[31] = success as u8;
```
Comment (belfortep) [inline src/precompiles.rs:290]: I like this way of returning the final value

**Review by belfortep — COMMENTED** [review]
Just asked for some clarification, the rest LGTM!

File: src/precompiles.rs (hunk 2 above)
Comment (juanimedone) [reply src/precompiles.rs:166]: Yes, it's just a naming thing. Both names refer to the same curve.

File: src/precompiles.rs (hunk 3 above)
Comment (juanimedone) [reply src/precompiles.rs]: Done!

**Review by belfortep — APPROVED** [review]

**Review by TomasArrachea — COMMENTED** [review]
just some style suggestions

File: src/precompiles.rs [hunk 5]
```diff
@@ -124,6 +142,156 @@ pub fn modexp(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes
...
+        }
+
+        let p1 = match BN254Curve::create_point_from_affine(g1_x, g1_y) {
+            Ok(point) => point,
+            Err(_) => return Bytes::from([0u8; 32].to_vec()),
+        };
+        let p2 = match BN254TwistCurve::create_point_from_affine(g2_x, g2_y) {
+            Ok(point) => point,
+            Err(_) => return Bytes::from([0u8; 32].to_vec()),
+        };
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply src/precompiles.rs:277]: Yes, that looks way nicer.

File: src/precompiles.rs [hunk 6]
```diff
@@ -124,6 +142,156 @@ pub fn modexp(calldata: &Bytes, gas_limit: u64, consumed_gas: &mut u64) -> Bytes
...
+
+        let pairing_result = match BN254AtePairing::compute_batch(&[(&p1, &p2)]) {
+            Ok(result) => result,
+            Err(_) => return Bytes::from([0u8; 32].to_vec()),
+        };
+        mul *= pairing_result;
+    }
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply src/precompiles.rs]: Done!

**Review by MegaRedHand — APPROVED** [review]
🚀 🚀 🚀
